### PR TITLE
implement doubly_lexical_ordering of 01-matrix in ```src/sage/matrix/matrix_mod2_dense.pyx```

### PR DESCRIPTION
**Pull request by kappybar**
Implement a doubly lexical ordering of a 01-matrix.  This is the first step to detect a graph is chordal bipartite or not as in #38792 (We can detect a graph is chordal bipartite or not by getting a doubly lexical ordering of its bipartite adjacency matrix and check it is $\Gamma$-free). 

The algorithm is introduced in [1]. There are other faster algorithms but it is the simplest one as far as I found. The time complexity is $O(nm^2)$ for $n \times m$ 01-matrix.

[1] Hoffman, Alan J., Anthonius Wilhelmus Johannes Kolen, and Michel Sakarovitch. "Totally-balanced and greedy matrices." SIAM Journal on Algebraic Discrete Methods 6.4 (1985): 721-730.

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

### CODE REVIEW

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 1]
```diff
@@ -2132,6 +2132,169 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+
+        ALGORITHM:
+
+        The algorithm is adapted from section 3 of [HAM1985]_. The time
+        complexity of this algorithm is `O(n \cdot m^2)` for a `n \times m`
+        matrix.
+
+        EXAMPLES::
+
+            sage: A = Matrix(GF(2), [
```
Comment (dcoudert) [inline src/sage/matrix/matrix_mod2_dense.pyx]: Why not using
```sage
            sage: A = Matrix(GF(2), [[0, 1],
            ....:                    [1, 0]])
```

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 2]
```diff
@@ -2132,6 +2132,169 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+                parition_i = 0
+                for row in reversed(range(A._nrows)):
+                    count1[col][parition_i] += 1 if mzd_read_bit(A._entries, row, col) else 0
+                    if row > 0 and partition_rows[row - 1]:
+                        parition_i += 1
+
+            # calculate largest_col = col s.t. count1[col] is lexicographically largest (0 <= col < i)
+            largest_col = 0
+            largest_count1 = count1[0]
+            for col in range(1, i):
```
Comment (dcoudert) [inline src/sage/matrix/matrix_mod2_dense.pyx]: if I understand well you want `largest_count1, largest_col = max((c, i) for c, I in enumerate(count1))` (in case of equality you to take the right most).
I don't know which form is best.

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 3]
```diff
@@ -2132,6 +2132,169 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+        cdef Matrix_mod2_dense A = self if inplace else self.__copy__()
+
+        for i in reversed(range(1, A._ncols + 1)):
+
+            # count 1 for each partition and column
+            count1 = [[0 for _ in range(partition_num)] for _ in range(i)]
+            for col in range(i):
+                parition_i = 0
+                for row in reversed(range(A._nrows)):
+                    count1[col][parition_i] += 1 if mzd_read_bit(A._entries, row, col) else 0
```
Comment (dcoudert) [inline src/sage/matrix/matrix_mod2_dense.pyx]: method `mzd_read_bit` returns a bit, right ? so you can directly add it without if, no ?

**Review by dcoudert — CHANGES_REQUESTED** [review]
I'm not an expert in this kind of methods, so I don't know if it can be improved further.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 1 above)
Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx]: Thank you for your review. I fixed it.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 2 above)
Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx]: I changed it because it is simpler. Note that the right most is not important in this algorithm. We can take any maximum element in case of equality.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 3 above)
Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx]: I confirmed ```mzd_read_bit``` returns ```int```, so I changed it. Thank you for your review.

**Review by dcoudert — COMMENTED** [review]
You you raise an error when `inplece` is True and the matrix is immutable.
`A = Matrix(GF(2), [[0, 1], [1, 0]], immutable=True)`

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 4]
```diff
@@ -2132,6 +2132,173 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+
+            sage: A = Matrix(GF(2), [[0, 1],
+            ....:                    [1, 0]])
+            sage: r, c = A.doubly_lexical_ordering(inplace=True); A
+            [1 0]
+            [0 1]
+
+        TESTS:
+
+            sage: A = Matrix(GF(2), [[1, 1, 0, 0, 0, 0, 0],
```
Comment (dcoudert) [inline src/sage/matrix/matrix_mod2_dense.pyx:2196]: can you add an explanation of what this test is about ?
Furthermore, we need something ending with `::` before this doctest.

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 5]
```diff
@@ -2132,6 +2132,173 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+            sage: flag = True
+            sage: for i in range(B.ncols()):
+            ....:     for j in range(i):
+            ....:         for k in reversed(range(B.nrows())):
+            ....:             if B[k][j] > B[k][i]:
+            ....:                 flag = False
+            ....:                 break
+            ....:             if B[k][j] < B[k][i]:
+            ....:                 break
+            ....:
```
Comment (dcoudert) [inline src/sage/matrix/matrix_mod2_dense.pyx]: this line is not needed.

Also, you should check the value of `flag` before starting the next block of for loops.

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 6]
```diff
@@ -2132,6 +2132,173 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+        An immutable matrix calling with ``inplace=True`` will raise an error::
+
+            sage: A = Matrix(GF(2), [[0, 1], [1, 0]], immutable=True)
+            sage: r, c = A.doubly_lexical_ordering(inplace=True)
+            Traceback (most recent call last):
+            ...
+            TypeError: This matrix is immutable and can thus not be changed. Use inplace=False or create a mutable copy.
+
+        """
+
```
Comment (dcoudert) [inline src/sage/matrix/matrix_mod2_dense.pyx:2233]: no need for this empty line

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 7]
```diff
@@ -2132,6 +2132,173 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+            sage: A == B
+            True
+
+        An immutable matrix calling with ``inplace=True`` will raise an error::
+
+            sage: A = Matrix(GF(2), [[0, 1], [1, 0]], immutable=True)
+            sage: r, c = A.doubly_lexical_ordering(inplace=True)
+            Traceback (most recent call last):
+            ...
+            TypeError: This matrix is immutable and can thus not be changed. Use inplace=False or create a mutable copy.
```
Comment (dcoudert) [inline src/sage/matrix/matrix_mod2_dense.pyx]: Could use the following form to avoid a long line.
```sage
            TypeError: This matrix is immutable and can thus not be changed.
             Use inplace=False or create a mutable copy.
```

**Review by dcoudert — CHANGES_REQUESTED** [review]
it would be nice if someone could launch CI tests.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 4 above)
Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx:2196]: I added an explanation of this test. I have no reason to choose this specific matrix, but I added this test to confirm the correctness of this implementation.

**Review by dcoudert — APPROVED** [review]
LGTM.

**Comment by dcoudert** [comment]
@tscrim, before I set this PR to positive review, could you have a look to check that it's ok. Furthermore, can you launch the CIs ? Thanks.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 6 above)
Comment (tscrim) [reply src/sage/matrix/matrix_mod2_dense.pyx:2233]: This still has not been addressed (in fact, our general style does not have it).

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 8]
```diff
@@ -2132,6 +2132,176 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+            sage: r, c = A.doubly_lexical_ordering(inplace=True)
+            Traceback (most recent call last):
+            ...
+            TypeError: This matrix is immutable and can thus not be changed.
+             Use inplace=False or create a mutable copy.
+        """
+
+        if inplace and self.is_immutable():
+            raise TypeError("This matrix is immutable and can thus not be changed."
+                            " Use inplace=False or create a mutable copy.")
```
Comment (tscrim) [inline src/sage/matrix/matrix_mod2_dense.pyx]:
```suggestion
            raise TypeError("this matrix is immutable;"
                            " use inplace=False or apply to a mutable copy")
```
This follows Python's error message conventions (which we generally follow).

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 9]
```diff
@@ -2132,6 +2132,176 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+        A lexical ordering of vectors is the standard dictionary ordering,
+        except that vectors will be read from highest to lowest coordinate.
+        Thus row vectors will be compared from right to left, and column
+        vectors from bottom to top.
+
+        INPUT:
+
+        - ``inplace`` -- boolean (default: ``False``); using ``inplace=True``
+          will permute the rows and columns of the current matrix
+          according to a doubly lexical ordering. This will modify the matrix.
```
Comment (tscrim) [inline src/sage/matrix/matrix_mod2_dense.pyx]:
```suggestion
        - ``inplace`` -- boolean (default: ``False``); using ``inplace=True``
          will permute the rows and columns of the current matrix
          according to a doubly lexical ordering; this will modify the matrix
```

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 10]
```diff
@@ -2132,6 +2132,176 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+
+        OUTPUT:
+
+        Returns a pair (``row_ordering``, ``col_ordering``). Each item
+        is a ``PermutationGroupElement`` that represents a doubly lexical
+        ordering of the rows or columns.
```
Comment (tscrim) [inline src/sage/matrix/matrix_mod2_dense.pyx]:
```suggestion
        A pair ``(row_ordering, col_ordering)`` of
        :class:`~sage.groups.perm_gps.constructor.PermutationGroupElement`
        that represents a doubly lexical ordering of the rows or columns.
```

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 11]
```diff
@@ -2132,6 +2132,176 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+
+        .. SEEALSO::
+
+            - :meth:`~sage.matrix.matrix2.Matrix.permutation_normal_form` --
+              a similar matrix normal form
```
Comment (tscrim) [inline src/sage/matrix/matrix_mod2_dense.pyx]:
```suggestion
            :meth:`~sage.matrix.matrix2.Matrix.permutation_normal_form`;
            a similar matrix normal form
```

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 12]
```diff
@@ -2132,6 +2132,176 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+            sage: for i in range(B.ncols()):
+            ....:     for j in range(i):
+            ....:         for k in reversed(range(B.nrows())):
+            ....:             if B[k][j] > B[k][i]:
+            ....:                 flag = False
+            ....:                 break
+            ....:             if B[k][j] < B[k][i]:
+            ....:                 break
+            sage: flag
+            True
```
Comment (tscrim) [inline src/sage/matrix/matrix_mod2_dense.pyx]:
```suggestion
            ....:             assert B[k][j] <= B[k][i]
            ....:             if B[k][j] < B[k][i]:
            ....:                 break
```
I think this is easier to understand and will short-circuit faster. Please also make a similar change below.

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 13]
```diff
@@ -2132,6 +2132,176 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+        partition_num = 1
+        row_swapped = list(range(1, self._nrows + 1))
+        col_swapped = list(range(1, self._ncols + 1))
+
+        cdef Matrix_mod2_dense A = self if inplace else self.__copy__()
+
+        for i in reversed(range(1, A._ncols + 1)):
+
+            # count 1 for each partition and column
+            count1 = [[0 for _ in range(partition_num)] for _ in range(i)]
```
Comment (tscrim) [inline src/sage/matrix/matrix_mod2_dense.pyx]:
```suggestion
            count1 = [[0]*partition_num for _ in range(i)]
```

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 14]
```diff
@@ -2132,6 +2132,176 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+        partition_rows = [False for _ in range(self._nrows - 1)]
+        partition_num = 1
+        row_swapped = list(range(1, self._nrows + 1))
+        col_swapped = list(range(1, self._ncols + 1))
+
+        cdef Matrix_mod2_dense A = self if inplace else self.__copy__()
+
+        for i in reversed(range(1, A._ncols + 1)):
+
+            # count 1 for each partition and column
```
Comment (tscrim) [inline src/sage/matrix/matrix_mod2_dense.pyx]:
```suggestion
        cdef int i, col, partition_i, partition_start, partition_end
        cdef int row_start, row_end
        for i in reversed(range(1, A._ncols + 1)):
            # count 1 for each partition and column
```

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 15]
```diff
@@ -2132,6 +2132,176 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+            count1 = [[0 for _ in range(partition_num)] for _ in range(i)]
+            for col in range(i):
+                parition_i = 0
+                for row in reversed(range(A._nrows)):
+                    count1[col][parition_i] += mzd_read_bit(A._entries, row, col)
+                    if row > 0 and partition_rows[row - 1]:
+                        parition_i += 1
```
Comment (tscrim) [inline src/sage/matrix/matrix_mod2_dense.pyx]:
```suggestion
                for row in reversed(range(1, A._nrows)):
                    count1[col][parition_i] += mzd_read_bit(A._entries, row, col)
                    if partition_rows[row - 1]:
                        parition_i += 1
                count1[col][parition_i] += mzd_read_bit(A._entries, 0, col)  # special case of row == 0
```

File: src/sage/matrix/matrix_mod2_dense.pyx [hunk 16]
```diff
@@ -2132,6 +2132,176 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
...
+        """
+
+        if inplace and self.is_immutable():
+            raise TypeError("This matrix is immutable and can thus not be changed."
+                            " Use inplace=False or create a mutable copy.")
+
+        partition_rows = [False for _ in range(self._nrows - 1)]
+        partition_num = 1
+        row_swapped = list(range(1, self._nrows + 1))
+        col_swapped = list(range(1, self._ncols + 1))
```
Comment (tscrim) [inline src/sage/matrix/matrix_mod2_dense.pyx]:
```suggestion
        cdef list partition_rows = [False for _ in range(self._nrows - 1)]
        cdef int partition_num = 1
        cdef list row_swapped = list(range(1, self._nrows + 1))
        cdef list col_swapped = list(range(1, self._ncols + 1))
```

**Review by tscrim — CHANGES_REQUESTED** [review]
Here are some various changes that I would like incorporated before a positive review. Some of them of documentation, some of them are trying to get some cheap speed improvements by strongly typing variables for the Cython compiler.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 15 above)
Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx]: This change will cause segmentation fault when ``nrow == 0`` as in an example below.(but I am wondering the size (0, 4) of a matrix is valid or not.)

I think the current implementation is ok because the case ``row=0`` is already handled by ``if row > 0`` statement. 


```
sage: A = Matrix(GF(2), 0, 4, [])
sage: A
[]
sage: A.nrows()
0
sage: A.ncols()
4
sage: r, c = A.doubly_lexical_ordering()
------------------------------------------------------------------------
(no backtrace available)
------------------------------------------------------------------------
Unhandled SIGSEGV: A segmentation fault occurred.
This probably occurred because a *compiled* module has a bug
in it and is not properly wrapped with sig_on(), sig_off().
Python will now terminate.
------------------------------------------------------------------------
```

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 6 above)
Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx:2233]: I think this empty line is already erased.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 8 above)
Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx]: Thank you for suggestion! I fixed it

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 9 above)
Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx]: Thank you for suggestion! I fixed it

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 12 above)
Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx]: Thank you for suggestion! I fixed it

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 15 above)
Comment (tscrim) [reply src/sage/matrix/matrix_mod2_dense.pyx]: This is a good corner case check. Good catch too. These cases should be handled separately at the beginning of the method.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 6 above)
Comment (tscrim) [reply src/sage/matrix/matrix_mod2_dense.pyx:2233]: It's still there in the online diff though.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 15 above)
Comment (tscrim) [reply src/sage/matrix/matrix_mod2_dense.pyx]: By the first sentence, I mean add it to the `TESTS::` block in the docstring.

Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx]: I added tests and code to handle this corner case.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 6 above)
Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx:2233]: In my recent commit https://github.com/sagemath/sage/pull/39794/commits/c70956c4e20d9de4d9fb33b3b9a3dc575296155e , I removed an empty line around the end of tests. I am wondering why this online diff does not become outdated

Comment (tscrim) [reply src/sage/matrix/matrix_mod2_dense.pyx:2233]: Well, it is now not showing up in the online viewer either. Anyways, thanks.

File: src/sage/matrix/matrix_mod2_dense.pyx (hunk 15 above)
Comment (tscrim) [reply src/sage/matrix/matrix_mod2_dense.pyx]: Now my suggested change shouldn't seg fault. I expect it to be (marginally) faster.

Comment (kappybar) [reply src/sage/matrix/matrix_mod2_dense.pyx]: I agree with you. I changed.

**Review by tscrim — APPROVED** [review]
Thank you. Positive review from me now too.
